### PR TITLE
 Remove ad-hoc `Anticipate` objects

### PR DESCRIPTION
**Pull request by maximedenes**
We drop support for `Require` inside modules. Was already deprecated.

### CODE REVIEW

**Review by ejgallego — APPROVED** [review]
LGTM, I guess the impact on users shall be minimal, but I'm always afraid of how users doing very costly proofs may be impacted.

File: library/library.ml [hunk 1]
```diff
@@ -427,61 +427,20 @@ let register_library m =
...
-        warn_require_in_module ();
-	add_anonymous_leaf (in_require (needed,modrefl,None));
-	Option.iter (fun exp ->
-	  add_anonymous_leaf (in_import_library (modrefl,exp)))
-	  export
-      end
-    else
-      add_anonymous_leaf (in_require (needed,modrefl,export));
-  ()
+  if Lib.is_module_or_modtype () then
```
Comment (SkySkimmer) [inline library/library.ml]: The object also handled Require in Section, so either we should keep it around (but classified Dispose), or we should error when Require in Section.

Comment (maximedenes) [reply library/library.ml]: > The object also handled Require in Section, so either we should keep it around (but classified Dispose), or we should error when Require in Section.

Argh, good point. `Require` in section wasn't deprecated, was it?

Comment (maximedenes) [reply library/library.ml]: > `Require` in section wasn't deprecated, was it?

In fact, it was!

Comment (ejgallego) [reply library/library.ml]: It was however that may impact a few high-profile users :( :(

Comment (maximedenes) [reply library/library.ml]: Let's first evaluate what CI shows.

Comment (ejgallego) [reply library/library.ml]: I was thinking more of the use case Adam often reminds, like they are in the middle of a costly proof and they want to require some module without redoing everything. I dunno, just mentioning it out.

File: vernac/vernacentries.ml [hunk 2]
```diff
@@ -985,14 +985,11 @@ let vernac_end_segment ({v=id} as lid) =
...
-  let category = "deprecated" in
-  CWarnings.create ~name ~category
-    (fun () -> strbrk "Use of “Require” inside a section is deprecated.")
-
 let vernac_require from import qidl =
-  if Lib.sections_are_opened () then warn_require_in_section ();
+  if Lib.is_module_or_modtype () then
+    CErrors.user_err Pp.(str "The \"Require\" command is not supported inside modules or module types.");
+  if Lib.sections_are_opened () then
+    CErrors.user_err Pp.(str "The \"Require\" command is not supported inside sections.");
```
Comment (SkySkimmer) [inline vernac/vernacentries.ml]: It doesn't matter much but I think I'd like the errors better in library.ml rather than here.

**Comment by Zimmi48** [comment]
Dear PR author(s) and contributors, it seems to us this PR may need additional work beyond the 8.12 branching deadline on May 15th.

We are thus removing the 8.12+beta1 milestone; please retarget as appropriate.

We apologize in advance if we misunderstood the PR status.

**Review by JasonGross — CHANGES_REQUESTED** [review]
> I was thinking more of the use case Adam often reminds, like they are in the middle of a costly proof and they want to require some module without redoing everything. I dunno, just mentioning it out.

Indeed, please don't remove support for this.  I don't use it in any completed developments, but it's quite useful in the middle of developing slow proofs.

**Comment by ppedrot** [comment]
So I guess we're stuck to statu quo?

**Comment by SkySkimmer** [comment]
I don't believe we should remove support for Require in Sections/Modules. Since this PR is dead let's just close.

**Pull request closed without merging**